File: C++/RazorAHRS.cpp
/******************************************************************************************
* Mac OSX / Unix / Linux C++ Interface for Razor AHRS v1.4.2
* 9 Degree of Measurement Attitude and Heading Reference System
* for Sparkfun "9DOF Razor IMU" and "9DOF Sensor Stick"
*
* Released under GNU GPL (General Public License) v3.0
* Copyright (C) 2013 Peter Bartz [http://ptrbrtz.net]
* Copyright (C) 2011-2012 Quality & Usability Lab, Deutsche Telekom Laboratories, TU Berlin
* Written by Peter Bartz (peter-bartz@gmx.de)
*
* Infos, updates, bug reports,contributions and feedback:
*     https://github.com/ptrbrtz/razor-9dof-ahrs
******************************************************************************************/

#include "RazorAHRS.h"
#include <cassert>

RazorAHRS::RazorAHRS(const std::string &port, DataCallbackFunc data_func, ErrorCallbackFunc error_func,
    Mode mode, int connect_timeout_ms, speed_t speed)
    : _mode(mode)
    , _input_pos(0)
    , _connect_timeout_ms(connect_timeout_ms)
    , data(data_func)
    , error(error_func)
    , _thread_id(0)
    , _stop_thread(false)
{  
  // check data type sizes
  assert(sizeof(char) == 1);
  assert(sizeof(float) == 4);

  // open serial port
  if (port == "")
    throw std::runtime_error("No port specified!");
  if (!_open_serial_port(port.c_str()))
    throw std::runtime_error("Could not open serial port!");  
      
  // get port attributes
  struct termios tio;
  if (int errorID = tcgetattr(_serial_port, &tio))
    throw std::runtime_error("Could not get serial port attributes! Error # " + to_str(errorID));
  
  /* see http://www.easysw.com/~mike/serial/serial.html */
  /* and also http://linux.die.net/man/3/tcsetattr */
  // basic raw/non-canonical setup
  cfmakeraw(&tio);
  
  // enable reading and ignore control lines
  tio.c_cflag |= CREAD | CLOCAL;

  // set 8N1
  tio.c_cflag &= ~PARENB; // no parity bit
  tio.c_cflag &= ~CSTOPB; // only one stop bit
  tio.c_cflag &= ~CSIZE;  // clear data bit number
  tio.c_cflag |= CS8;     // set 8 data bits
  
  // no hardware flow control
  tio.c_cflag &= ~CRTSCTS;
  // no software flow control  
  tio.c_iflag &= ~(IXON | IXOFF | IXANY);
  
  // poll() is broken on OSX, so we set VTIME and use read(), which is ok since
  // we're reading only one port anyway
  tio.c_cc[VMIN]  = 0;
  tio.c_cc[VTIME] = 10; // 10 * 100ms = 1s
  
  // set port speed
  if (int errorID = cfsetispeed(&tio, speed))
    throw std::runtime_error(" " + to_str(errorID)
        + ": Could not set new serial port input speed to "
        + to_str(speed) + ".");
  if (int errorID = cfsetospeed(&tio, speed))
    throw std::runtime_error(" " + to_str(errorID)
        + ": Could not set new serial port output speed to "
        + to_str(speed) + ".");

  // set port attributes
  // must be done after setting speed!
  if (int errorID = tcsetattr(_serial_port, TCSANOW, &tio))
  {
    throw std::runtime_error(" " + to_str(errorID)
        + ": Could not set new serial port attributes.");
  }

  // start input/output thread
  _start_io_thread();
}

RazorAHRS::~RazorAHRS()
{
  // if thread was started, stop thread
  if (_thread_id) _stop_io_thread();
  close(_serial_port);
}

bool
RazorAHRS::_read_token(const std::string &token, char c)
{
  if (c == token[_input_pos++])
  {
    if (_input_pos == token.length())
    {
      // synch token found
      _input_pos = 0;
      return true;
    }
  }
  else
  {
    _input_pos = 0;
  }
  
  return false;
}

bool
RazorAHRS::_init_razor()
{
  char in;
  int result;
  struct timeval t0, t1, t2;
  const std::string synch_token = "#SYNCH";
  const std::string new_line = "\r\n";

  // start time
  gettimeofday(&t0, NULL);

  // request synch token to see if Razor is really present
  const std::string contact_synch_id = "00"; 
  const std::string contact_synch_request = "#s" + contact_synch_id; 
  const std::string contact_synch_reply = synch_token + contact_synch_id + new_line;
  write(_serial_port, contact_synch_request.data(), contact_synch_request.length());
  gettimeofday(&t1, NULL);

  // set non-blocking I/O
  if (!_set_nonblocking_io()) return false;

  /* look for tracker */
  while (true)
  {
    // try to read one byte from the port
    result = read(_serial_port, &in, 1);
    
    // one byte read
    if (result > 0)
    {
      if (_read_token(contact_synch_reply, in))
        break;
    }
    // no data available
    else if (result == 0)
      usleep(1000); // sleep 1ms
    // error?
    else
    {
      if (errno != EAGAIN && errno != EINTR)
        throw std::runtime_error("Can not read from serial port (1).");
    }

    // check timeout
    gettimeofday(&t2, NULL);
    if (elapsed_ms(t1, t2) > 200)
    {
      // 200ms elapsed since last request and no answer -> request synch again
      // (this happens when DTR is connected and Razor resets on connect)
      write(_serial_port, contact_synch_request.data(), contact_synch_request.length());
      t1 = t2;
    }
    if (elapsed_ms(t0, t2) > _connect_timeout_ms)
      // timeout -> tracker not present
      throw std::runtime_error("Can not init: tracker does not answer.");
  }
  
  
  /* configure tracker */
  // set correct binary output mode, enable continuous streaming, disable errors and
  // request synch token. So we're good, no matter what state the tracker
  // currently is in.
  const std::string config_synch_id = "01";
  const std::string config_synch_reply = synch_token + config_synch_id + new_line;

  std::string config = "#o1#oe0#s" + config_synch_id;
  if (_mode == YAW_PITCH_ROLL) config = "#ob" + config;
  else if (_mode == ACC_MAG_GYR_RAW) config = "#osrb" + config;
  else if (_mode == ACC_MAG_GYR_CALIBRATED) config = "#oscb" + config;
  else throw std::runtime_error("Can not init: unknown 'mode' parameter.");  

  write(_serial_port, config.data(), config.length());
  
  // set blocking I/O
  // (actually semi-blocking, because VTIME is set)
  if (!_set_blocking_io()) return false;

  while (true)
  {    
    // try to read one byte from the port
    result = read(_serial_port, &in, 1);
    
    // one byte read
    if (result > 0)
    {
      if (_read_token(config_synch_reply, in))
        break;  // alrighty
    }
    // error?
    else
    {
      if (errno != EAGAIN && errno != EINTR)
        throw std::runtime_error("Can not read from serial port (2).");
    }
  }
  
  // we keep using blocking I/O
  //if (_set_blocking_io() == -1)
  //  return false;
  
  return true;
}

bool
RazorAHRS::_open_serial_port(const char *port)
{
  // O_NDELAY allows open even with no carrier detect (e.g. needed for Razor)
  if ((_serial_port = open(port, O_RDWR | O_NOCTTY | O_NDELAY)) != -1)
  {
    // make I/O blocking again
    if (_set_blocking_io()) return true;
  }
  
  // something didn't work
  close(_serial_port);
  return false;
}

bool
RazorAHRS::_set_blocking_io()
{
  int flags;
  
  // clear O_NDELAY to make I/O blocking again
  // in fact this is semi-blocking, since we set VTIME on the port
  if (((flags = fcntl(_serial_port, F_GETFL, 0)) != -1) &&
      (fcntl(_serial_port, F_SETFL, flags & ~O_NDELAY)) != -1)
  {
    return true;
  }
  
  return false;
}

bool
RazorAHRS::_set_nonblocking_io()
{
  int flags;

  // set O_NDELAY to make I/O non-blocking
  if (((flags = fcntl(_serial_port, F_GETFL, 0)) != -1) &&
      (fcntl(_serial_port, F_SETFL, flags | O_NDELAY)) != -1)
  {
    return true;
  }
  
  return false;
}

bool
RazorAHRS::_is_io_blocking()
{
  return (fcntl(_serial_port, F_GETFL, 0) & O_NDELAY);
}

void*
RazorAHRS::_thread(void *arg)
{
  char c;
  int result;
  
  try
  {
    if (!_init_razor())
    {
      error("Tracker init failed.");
      return arg;
    }
  }
  catch(std::runtime_error& e)
  {
    error("Tracker init failed: " + std::string(e.what()));
    return arg;
  }
  
  while (!_stop_thread)
  {
    if ((result = read(_serial_port, &c, 1)) > 0) // blocks only for VTIME before returning
    {
      // read binary stream
      // (type-punning: aliasing with char* is ok)
      (reinterpret_cast<char*> (&_input_buf))[_input_pos++] = c;
      
      if (_mode == YAW_PITCH_ROLL) {  // 3 floats
        if (_input_pos == 12) // we received a full frame
        {
          // convert endianess if necessary
          if (_big_endian())
          {
            _swap_endianess(_input_buf.ypr, 3);
          }
          
          // invoke callback
          data(_input_buf.ypr);
          
          _input_pos = 0;
        }
      } else { // raw or calibrated sensor data (9 floats)
        if (_input_pos == 36) // we received a full frame
        {
          // convert endianess if necessary
          if (_big_endian())
          {
            _swap_endianess(_input_buf.amg, 9);
          }
          
          // invoke callback
          data(_input_buf.amg);
          
          _input_pos = 0;
        }
      }
    }
    // error?
    else if (result < 0)
    {
      if (errno != EAGAIN && errno != EINTR)
      {
        error("Can not read from serial port (3).");
        return arg;
      }
    }
    // else if result is 0, no data was available
  }

  return arg;
}

File: Android/RazorAHRS/src/de/tuberlin/qu/razorahrs/RazorAHRS.java
/******************************************************************************************
* Android Java Interface for Razor AHRS v1.4.2
* 9 Degree of Measurement Attitude and Heading Reference System
* for Sparkfun "9DOF Razor IMU" and "9DOF Sensor Stick"
*
* Released under GNU GPL (General Public License) v3.0
* Copyright (C) 2013 Peter Bartz [http://ptrbrtz.net]
* Copyright (C) 2011-2012 Quality & Usability Lab, Deutsche Telekom Laboratories, TU Berlin
* Written by Peter Bartz (peter-bartz@gmx.de)
*
* Infos, updates, bug reports, contributions and feedback:
*     https://github.com/ptrbrtz/razor-9dof-ahrs
******************************************************************************************/

package de.tuberlin.qu.razorahrs;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.UUID;

import org.apache.http.util.EncodingUtils;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.os.Handler;
import android.os.Message;
import android.os.SystemClock;
import android.util.Log;

/**
 * Class to easily interface the Razor AHRS via Bluetooth.
 * <p>
 * Bluetooth seems to be even more picky on Android than it is anyway. Be sure to have a look at the
 * section about Android Bluetooth in the tutorial at
 * <a href="https://github.com/ptrbrtz/razor-9dof-ahrs">
 * https://github.com/ptrbrtz/razor-9dof-ahrs</a>!
 * <p>
 * The app using this class has to
 * <ul>
 * <li>target Android 2.0 (API Level 5) or later.
 * <li>specify the uses-permissions <code>BLUETOOTH</code> and <code>BLUETOOTH_ADMIN</code> in it's
 * AndroidManifest.xml.
 * <li>add this Library Project as a referenced library (Project Properties -> Android -> Library)
 * </ul>
 * <p>
 * TODOs:
 * <ul>
 * <li>Add support for USB OTG (Android device used as USB host), if using FTDI is possible.
 * </ul>
 * 
 * @author Peter Bartz
 */
public class RazorAHRS {
	private static final String TAG = "RazorAHRS";
	private static final boolean DEBUG = false;
	private static final String SYNCH_TOKEN = "#SYNCH";
	private static final String NEW_LINE = "\r\n";

	// Timeout to init Razor AHRS after a Bluetooth connection has been established
	public static final int INIT_TIMEOUT_MS = 5000;

	// IDs passed to internal message handler
	private static final int MSG_ID__YPR_DATA = 0;
	private static final int MSG_ID__IO_EXCEPTION_AND_DISCONNECT = 1;
	private static final int MSG_ID__CONNECT_OK = 2;
	private static final int MSG_ID__CONNECT_FAIL = 3;
	private static final int MSG_ID__CONNECT_ATTEMPT = 4;
	private static final int MSG_ID__AMG_DATA = 5;

	private static final UUID UUID_SPP = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");

	/**
	 * Razor output modes.
	 * Use <code>YAW_PITCH_ROLL_ANGLES</code> to receive yaw, pitch and roll in degrees. <br>
	 * Use <code>RAW_SENSOR_DATA</code> or <code>CALIBRATED_SENSOR_DATA</code> to read raw or
	 * calibrated xyz sensor data of the accelerometer, magnetometer and the gyroscope.
	 */
	public enum RazorOutputMode {
		YAW_PITCH_ROLL_ANGLES,
		RAW_SENSOR_DATA,
		CALIBRATED_SENSOR_DATA
	}
	private RazorOutputMode razorOutputMode;

	private enum ConnectionState {
		DISCONNECTED,
		CONNECTING,
		CONNECTED,
		USER_DISCONNECT_REQUEST
	}
	volatile private ConnectionState connectionState = ConnectionState.DISCONNECTED;

	volatile private BluetoothSocket btSocket;
	volatile private BluetoothDevice btDevice;
	volatile private InputStream inStream;
	volatile private OutputStream outStream;

	private RazorListener razorListener;
	private boolean callbacksEnabled = true;

	BluetoothThread btThread;

	private int numConnectAttempts;

	// Object pools
	private ObjectPool<float[]> float3Pool = new ObjectPool<float[]>(new ObjectPool.ObjectFactory<float[]>() {
		@Override
		public float[] newObject() {
			return new float[3];
		}
	});
	private ObjectPool<float[]> float9Pool = new ObjectPool<float[]>(new ObjectPool.ObjectFactory<float[]>() {
		@Override
		public float[] newObject() {
			return new float[9];
		}
	});

	/**
	 * Constructor.
	 * Must be called from the thread where you want receive the RazorListener callbacks! So if you
	 * want to manipulate Android UI from the callbacks you have to call this from your main/UI
	 * thread.
	 *
	 * @param btDevice {@link android.bluetooth.BluetoothDevice BluetoothDevice} holding the Razor
	 * 		AHRS to connect to.
	 * @param razorListener {@link RazorListener} that will be notified of Razor AHRS events.
	 * @throws RuntimeException thrown if one of the parameters is null.
	 */
	public RazorAHRS(BluetoothDevice btDevice, RazorListener razorListener)
			throws RuntimeException {
		this(btDevice, razorListener, RazorOutputMode.YAW_PITCH_ROLL_ANGLES);
	}

	/**
	 * Constructor.
	 * Must be called from the thread where you want receive the RazorListener callbacks! So if you
	 * want to manipulate Android UI from the callbacks you have to call this from your main/UI
	 * thread.
	 *
	 * @param btDevice {@link android.bluetooth.BluetoothDevice BluetoothDevice} holding the Razor
	 * 		AHRS to connect to.
	 * @param razorListener {@link RazorListener} that will be notified of Razor AHRS events.
	 * @param razorOutputMode {@link RazorOutputMode} that you desire.
	 * @throws RuntimeException thrown if one of the parameters is null.
	 */
	public RazorAHRS(BluetoothDevice btDevice, RazorListener razorListener, RazorOutputMode razorOutputMode)
			throws RuntimeException {
		if (btDevice == null)
			throw new RuntimeException("BluetoothDevice can not be null.");
		this.btDevice = btDevice;

		if (razorListener == null)
			throw new RuntimeException("RazorListener can not be null.");
		this.razorListener = razorListener;

		if (razorOutputMode == null)
			throw new RuntimeException("RazorMode can not be null.");
		this.razorOutputMode = razorOutputMode;
	}

	/**
	 * @return <code>true</code> if listener callbacks are currently enabled, <code>false</code> else.
	 */
	public boolean getCallbacksEnabled() {
		return callbacksEnabled;
	}

	/**
	 * Enables/disables listener callbacks.
	 * @param enabled
	 */
	public void setCallbacksEnabled(boolean enabled) {
		callbacksEnabled = enabled;
	}

	/**
	 * Connect and start reading. Both is done asynchronously. {@link RazorListener#onConnectOk()}
	 * or {@link RazorListener#onConnectFail(IOException)} callbacks will be invoked.
	 * 
	 * @param numConnectAttempts Number of attempts to make when trying to connect. Often connecting
	 * 		only works on the 2rd try or later. Bluetooth hooray.
	 */
	public void asyncConnect(int numConnectAttempts) {
		if (DEBUG) Log.d(TAG, "asyncConnect() BEGIN");
		// Disconnect and wait for running thread to end, if needed
		if (btThread != null) {
			asyncDisconnect();
			try {
				btThread.join();
			} catch (InterruptedException e) { }
		}

		// Bluetooth thread not running any more, we're definitely in DISCONNECTED state now

		// Create new thread to connect to Razor AHRS and read input
		this.numConnectAttempts = numConnectAttempts;
		connectionState = ConnectionState.CONNECTING;
		btThread = new BluetoothThread();
		btThread.start();
		if (DEBUG) Log.d(TAG, "asyncConnect() END");
	}

	/**
	 * Disconnects from Razor AHRS. If still connecting this will also cancel the connection process.
	 */
	public void asyncDisconnect() {
		if (DEBUG) Log.d(TAG, "asyncDisconnect() BEGIN");
		synchronized (connectionState) {
			if (DEBUG) Log.d(TAG, "asyncDisconnect() SNYNCHRONIZED");
			// Don't go to USER_DISCONNECT_REQUEST state if we are disconnected already
			if (connectionState == ConnectionState.DISCONNECTED)
				return;

			// This is a wanted disconnect, so we force (blocking) I/O to break
			connectionState = ConnectionState.USER_DISCONNECT_REQUEST;
			closeSocketAndStreams();
		}
		if (DEBUG) Log.d(TAG, "asyncDisconnect() END");
	}

	/**
	 * Writes out a string using ASCII encoding. Assumes we're connected. Does not handle
	 * exceptions itself.
	 * 
	 * @param text Text to send out
	 * @throws IOException
	 */
	private void write(String text) throws IOException {
		outStream.write(EncodingUtils.getAsciiBytes(text));
	}

	/**
	 * Closes I/O streams and Bluetooth socket.
	 */
	private void closeSocketAndStreams() {
		if (DEBUG) Log.d(TAG, "closeSocketAndStreams() BEGIN");
		// Try to switch off streaming output of Razor in preparation of next connect
		try {
			if (outStream != null)
				write("#o0");
		} catch (IOException e) { }

		// Close Bluetooth socket => I/O operations immediately will throw exception
		try {
			if (btSocket != null)
				btSocket.close();
		} catch (IOException e) { }
		if (DEBUG) Log.d(TAG, "closeSocketAndStreams() BT SOCKET CLOSED");
		
		// Close streams
		try {
			if (inStream != null)
				inStream.close();
		} catch (IOException e) { }
		try {
			if (outStream != null)
				outStream.close();
		} catch (IOException e) { }
		if (DEBUG) Log.d(TAG, "closeSocketAndStreams() STREAMS CLOSED");

		// Do not set socket and streams null, because input thread might still access them
		//inStream = null;
		//outStream = null;
		//btSocket = null;
		
		if (DEBUG) Log.d(TAG, "closeSocketAndStreams() END");
	}

	/**
	 * Thread that handles connecting to and reading from Razor AHRS. 
	 */
	private class BluetoothThread extends Thread {
		byte[] inBuf = new byte[512];
		int inBufPos = 0;
		
		/**
		 * Blocks until it can read one byte of input, assumes we have a connection up and running.
		 * 
		 * @return One byte from input stream
		 * @throws IOException If reading input stream fails
		 */
		private byte readByte() throws IOException {
			int in = inStream.read();
			if (in == -1)
				throw new IOException("End of Stream");
			return (byte) in;
		}

		/**
		 * Converts a buffer of bytes to an array of floats. This method does not do any error
		 * checking on parameter array sizes.
		 * @param byteBuf Byte buffer with length of at least <code>numFloats  * 4</code>.
		 * @param floatArr Float array with length of at least <code>numFloats</code>.
		 * @param numFloats Number of floats to convert
		 */
		private void byteBufferToFloatArray(byte[] byteBuf, float[] floatArr, int numFloats) {
			//int numFloats = byteBuf.length / 4;
			for (int i = 0; i < numFloats * 4; i += 4) {
				// Convert from little endian (Razor) to big endian (Java) and interpret as float
				floatArr[i/4] = Float.intBitsToFloat((byteBuf[i] & 0xff) + ((byteBuf[i+1] & 0xff) << 8) +
						((byteBuf[i+2] & 0xff) << 16) + ((byteBuf[i+3] & 0xff) << 24));
			}
		}

		/**
		 * Parse input stream for given token.
		 * @param token Token to find
		 * @param in Next byte from input stream
		 * @return <code>true</code> if token was found
		 */
		private boolean readToken(byte[] token, byte in) {
			if (in == token[inBufPos++]) {
				if (inBufPos == token.length) {
					// Synch token found
					inBufPos = 0;
					if (DEBUG) Log.d(TAG, "Token found");
					return true;
				}
			} else {
				inBufPos = 0;
			}
			
			return false;
		}

		/**
		 * Synches with Razor AHRS and sets parameters.
		 * @throws IOException
		 */
		private void initRazor() throws IOException {
			long t0, t1, t2;

			// Start time
			t0 = SystemClock.uptimeMillis();

			/* See if Razor is there */
			// Request synch token to see when Razor is up and running
			final String contactSynchID = "00";
			final String contactSynchRequest = "#s" + contactSynchID;
			final byte[] contactSynchReply = EncodingUtils.getAsciiBytes(SYNCH_TOKEN + contactSynchID + NEW_LINE);
			write(contactSynchRequest);
			t1 = SystemClock.uptimeMillis();

			while (true) {
				// Check timeout
				t2 = SystemClock.uptimeMillis();
				if (t2 - t1 > 200) {
					// 200ms elapsed since last request and no answer -> request synch again.
					// (This happens when DTR is connected and Razor resets on connect)
					write(contactSynchRequest);
					t1 = t2;
				}
				if (t2 - t0 > INIT_TIMEOUT_MS)
					// Timeout -> tracker not present
					throw new IOException("Can not init Razor: response timeout");

				// See if we can read something
				if (inStream.available() > 0) {
					// Synch token found?
					if (readToken(contactSynchReply, readByte()))
						break;
				} else {
					// No data available, wait
					delay(5); // 5ms
				}
			}

			/* Configure tracker */
			// Set binary output mode, enable continuous streaming, disable errors and request synch
			// token. So we're good, no matter what state the tracker currently is in.
			final String configSynchID = "01";
			final byte[] configSynchReply = EncodingUtils.getAsciiBytes(SYNCH_TOKEN + configSynchID + NEW_LINE);
			write("#ob#o1#oe0#s" + configSynchID);
			while (!readToken(configSynchReply, readByte())) { }
		}

		/**
		 * Opens Bluetooth connection to Razor AHRS.
		 * @throws IOException
		 */
		private void connect() throws IOException {		
			// Create Bluetooth socket
			btSocket = btDevice.createRfcommSocketToServiceRecord(UUID_SPP);
			if (btSocket == null) {
				if (DEBUG) Log.d(TAG, "btSocket is null in connect()");
				throw new IOException("Could not create Bluetooth socket");
			}

			// This could be used to create the RFCOMM socekt on older Android devices where
			//createRfcommSocketToServiceRecord is not present yet.
			/*try {
			Method m = btDevice.getClass().getMethod("createRfcommSocket", new Class[] { int.class });
			btSocket = (BluetoothSocket) m.invoke(btDevice, Integer.valueOf(1));
			} catch (Exception e) {
				throw new IOException("Could not create Bluetooth socket using reflection");
			}*/

			// Connect socket to Razor AHRS
			if (DEBUG) Log.d(TAG, "Canceling bt discovery");
			BluetoothAdapter.getDefaultAdapter().cancelDiscovery();	// Recommended
			if (DEBUG) Log.d(TAG, "Trying to connect() btSocket");
			btSocket.connect();

			// Get the input and output streams
			if (DEBUG) Log.d(TAG, "Trying to create streams");
			inStream = btSocket.getInputStream();
			outStream = btSocket.getOutputStream();
			if (inStream == null || outStream == null) {
				if (DEBUG) Log.d(TAG, "Could not create I/O stream(s) in connect()");
				throw new IOException("Could not create I/O stream(s)");
			}
		}

		/**
		 * Bluetooth I/O thread entry method.
		 */
		public void run() {
			if (DEBUG) Log.d(TAG, "Bluetooth I/O thread started");
			try {
				// Check if btDevice is set
				if (btDevice == null) {
					if (DEBUG) Log.d(TAG, "btDevice is null in run()");
					throw new IOException("Bluetooth device is null");
				}

				// Make several attempts to connect
				int i = 1;
				while (true) {
					if (DEBUG) Log.d(TAG, "Connect attempt " + i + " of " + numConnectAttempts);
					sendToParentThread(MSG_ID__CONNECT_ATTEMPT, i);
					try {
						connect();
						break;	// Alrighty!
					} catch (IOException e) {
						if (DEBUG) Log.d(TAG, "Attempt failed: " + e.getMessage());
						// Maximum number of attempts reached or cancel requested?
						if (i == numConnectAttempts || connectionState == ConnectionState.USER_DISCONNECT_REQUEST)
							throw e;

						// We couldn't connect on first try, manually starting Bluetooth discovery
						// often helps
						if (DEBUG) Log.d(TAG, "Starting BT discovery");
						BluetoothAdapter.getDefaultAdapter().startDiscovery();
						delay(5000); // 5 seconds - long enough?

						i++;
					}
				}

				// Set Razor output mode
				if (DEBUG) Log.d(TAG, "Trying to set Razor output mode");
				initRazor();

				// We're connected and initialized (unless disconnect was requested)
				synchronized (connectionState) {
					if (connectionState == ConnectionState.USER_DISCONNECT_REQUEST) {
						closeSocketAndStreams();
						throw new IOException(); // Dummy exception to force disconnect
					}
					else connectionState = ConnectionState.CONNECTED;
				}

				// Tell listener we're ready
				sendToParentThread(MSG_ID__CONNECT_OK, null);

				// Keep reading inStream until an exception occurs
				if (DEBUG) Log.d(TAG, "Starting input loop");
				while (true) {
					// Read byte from input stream
					inBuf[inBufPos++] = (byte) readByte();

					if (razorOutputMode == RazorOutputMode.YAW_PITCH_ROLL_ANGLES) {
						if (inBufPos == 12) {	// We received a full frame
							float[] ypr = float3Pool.get();
							byteBufferToFloatArray(inBuf, ypr, 3);

							// Forward to parent thread handler
							sendToParentThread(MSG_ID__YPR_DATA, ypr);

							// Rewind input buffer position
							inBufPos = 0;
						}
					} else {	// Raw or calibrated sensor data mode
						if (inBufPos == 36) {	// We received a full frame
							float[] amg = float9Pool.get();
							byteBufferToFloatArray(inBuf, amg, 9);

							// Forward to parent thread handler
							sendToParentThread(MSG_ID__AMG_DATA, amg);

							// Rewind input buffer position
							inBufPos = 0;
						}
					}
				}
			} catch (IOException e) {
				if (DEBUG) Log.d(TAG, "IOException in Bluetooth thread: " + e.getMessage());
				synchronized (connectionState) {
					// Don't forward exception if it was thrown because we broke I/O on purpose in
					// other thread when user requested disconnect
					if (connectionState != ConnectionState.USER_DISCONNECT_REQUEST) {
						// There was a true I/O error, cleanup and forward exception
						closeSocketAndStreams();
						if (DEBUG) Log.d(TAG, "Forwarding exception");
						if (connectionState == ConnectionState.CONNECTING)
							sendToParentThread(MSG_ID__CONNECT_FAIL, e);
						else
							sendToParentThread(MSG_ID__IO_EXCEPTION_AND_DISCONNECT, e);
					} else {
						// I/O error was caused on purpose, socket and streams are closed already
					}

					// I/O closed, thread done => we're disconnected now
					connectionState = ConnectionState.DISCONNECTED;
				}
			}
		}

		/**
		 * Sends a message to Handler assigned to parent thread.
		 * 
		 * @param msgId
		 * @param data
		 */
		private void sendToParentThread(int msgId, Object o) {
			if (callbacksEnabled)
				parentThreadHandler.obtainMessage(msgId, o).sendToTarget();
		}

		/**
		 * Sends a message to Handler assigned to parent thread.
		 * 
		 * @param msgId
		 * @param data
		 */
		private void sendToParentThread(int msgId, int i) {
			if (callbacksEnabled)
				parentThreadHandler.obtainMessage(msgId, i, -1).sendToTarget();
		}

		/**
		 * Wrapper for {@link Thread#sleep(long)};
		 * @param ms Milliseconds
		 */
		void delay(long ms) {
			try {
				sleep(ms); // Sleep 5ms
			} catch (InterruptedException e) { }
		}
	}

	/**
	 * Handler that forwards messages to the RazorListener callbacks. This handler runs in the
	 * thread this RazorAHRS object was created in and receives data from the Bluetooth I/O thread.
	 */
	private Handler parentThreadHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
				switch (msg.what) {
				case MSG_ID__YPR_DATA:	// Yaw, pitch and roll data
					float[] ypr = (float[]) msg.obj;
					razorListener.onAnglesUpdate(ypr[0], ypr[1], ypr[2]);
					float3Pool.put(ypr);
					break;
				case MSG_ID__AMG_DATA:	// Accelerometer, magnetometer and gyroscope data
					float[] amg = (float[]) msg.obj;
					razorListener.onSensorsUpdate(amg[0], amg[1], amg[2], amg[3], amg[4], amg[5],
							amg[6], amg[7], amg[8]);
					float9Pool.put(amg);
					break;
				case MSG_ID__IO_EXCEPTION_AND_DISCONNECT:
					razorListener.onIOExceptionAndDisconnect((IOException) msg.obj);
					break;
				case MSG_ID__CONNECT_ATTEMPT:
					razorListener.onConnectAttempt(msg.arg1, numConnectAttempts);
					break;
				case MSG_ID__CONNECT_OK:
					razorListener.onConnectOk();
					break;
				case MSG_ID__CONNECT_FAIL:
					razorListener.onConnectFail((IOException) msg.obj);
				break;
			}
		}
	};
}
